### DISCUSSION TITLE
getDefaultStore and testing without Provider

### DISCUSSION THREAD

**Discussion by ugurcanatas** [pmndrs/jotai#2747] [Q&A] [answered]
Hello good people of Jotai world

Recently we removed the Provider from our application. Doing so, some of the tests were started failing because the atom values are persisted(or previous value was kept is a better wording here) between the test blocks. 

We fixed the issue with broken tests by using getDefaultStore and reset necessary atoms to their initial values before each test.

```ts
const defaultStore = getDefaultStore()
defaultStore.set(someAtom, someInitialValue)
```

With that being said i have couple questions regarding this:

1 - When not using any Provider, does the approach that i mention above regarding resetting atoms to their initial state make sense? Is there any other approach that is a "better practice"?

2 - Is it possible to export the return type of `getDefaultStore`? which is `Store`. 

example usage of why:
```ts
// here the type is accessed via ReturnType and generic. Maybe its better with exported Store type?
export const _dev_resetSomeAtom = (defaultStore: ReturnType<typeof getDefaultStore>) => {
  defaultStore.set(someAtom, initialAtom)
}
```

**Comment by dai-shi** [comment, marked as answer]
2: We generally prefer `ReturnType<>` until we learn some unavoidable issues.

1: Resetting the default store would be challenging. I wonder if something like jest.mock can do it. Otherwise, you solution would be only the way. This shouldn't be Jotai-only problem. I wonder what's the best practice for resetting module variable in general in tests.

**Reply by ugurcanatas** [reply]
Yes, for sure it also relates to our internal setup, however we couldn't find the actual culprit yet.

Then we will keep the current "fix" as it is, unless we find a better way later on with jest.mock or something else.

Thanks for the response @dai-shi

**Reply by ugurcanatas** [reply]
Just as an update if it might help someone else, 

We wrote a small wrapper around atom(Primitive one). This adds the created atom to a Map and returns the created atom.

We loop over on this Map and reset all atoms to their initial state in setupTests. 

It's a more generic workaround at least but it still is a workaround.

This might be a way to go for anyone thats interested.

**Reply by dmaskasky** [reply]
Just a few words of caution with that approach,

keeping a reference to the atom around prevents it from being garbage collectable.

Don't forget to also add atomFamily atoms to your Map. You can use `family.getParams` and `family.unstable_listen` apis to get the atoms as they are created.

This approach will not work for any atoms created using the [atom factory pattern](https://jotai.org/docs/guides/composing-atoms#extending-atoms-with-atomwith-utils).

**Reply by ugurcanatas** [reply]
Oh okay, understood 🫡 

I'll bring this to my teammates attention.
Thanks a lot!
